scrapy_poet/utils.py

```diff
@@ -118,47 +120,57 @@ def is_response_going_to_be_used(request, spider):
     if is_callback_using_response(callback):
         return True
 
-    for provider in discover_callback_providers(callback):
+    providers = spider.settings["SCRAPY_POET_PROVIDERS"]
+    for provider in discover_callback_providers(callback, providers):
         if is_provider_using_response(provider):
             return True
 
     return False
 
 
-def build_plan(callback, instances) -> Tuple[andi.Plan, Dict[Type, PageObjectInputProvider]]:
+def build_plan(callback: Callable, providers: List[Type[PageObjectInputProvider]]) -> andi.Plan:
     """Build a plan for the injection in the callback."""
-    provider_instances = build_providers(instances)
-    plan = andi.plan(
+    return andi.plan(
         callback,
         is_injectable=is_injectable,
-        externally_provided=provider_instances.keys()
+        externally_provided=get_provided_classes_from_providers(providers),
```

Comment:
Initializing the providers from settings or getting the provided classes is something that is performed on each request and several times for each. It would be nice to cache these values so that they are initialized once at the start and then reusing. 

I propose to encapsulate all these functions (`build_plan`, `discovery_callback_providers`, maybe some of them present on the `InjectionMiddleware` itself, etc) into a class (e.g. `Injector`). This class is initialized in the Injection middleware. On initialization, it gets the settings, and caches some values (like the provider list and the list from `get_provided_classes_from_providers(providers)`) so that it can be reused.